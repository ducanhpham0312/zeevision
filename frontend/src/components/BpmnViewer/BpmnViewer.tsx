import NavigatedViewer from "bpmn-js/lib/NavigatedViewer";
import Viewer from "bpmn-js/lib/Viewer";
import { useCallback, useEffect, useRef, useState } from "react";
import { useElementSize } from "usehooks-ts";
import { Button } from "../Button";
import ZoomInIcon from "@mui/icons-material/ZoomIn";
import ZoomOutIcon from "@mui/icons-material/ZoomOut";
import CenterFocusWeakIcon from "@mui/icons-material/CenterFocusWeak";

// create a modeler

interface BpmnViewerProps {
  /**
   * The bpmn file content as string.
   */
  bpmnString: string;
  /**
   * The width (in pixels) that the Viewer should take.
   */
  width?: number;
  /**
   * The height (in pixels) that the Viewer should take.
   */
  height?: number;
  /**
   * Set this to true for NavigatedViewer which is interactable i.e. drag, zoom in out.
   */
  navigated?: boolean;

  control?: boolean;

  /**
   * List of bpmn element's intent for coloring BPMN diagram
   */
  colorOptions?: { elementId: string; intent: string }[];
}

type ViewBoxInner = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export function BpmnViewer({
  bpmnString,
  width,
  height,
  navigated,
  control,
  colorOptions,
}: BpmnViewerProps) {
  const containerRef = useRef<HTMLDivElement>(null);
  const [modeler, setModeler] = useState<Viewer>();
  const [currentZoom, setCurrentZoom] = useState(1);

  const zoomStrength = 0.3;

  useEffect(() => {
    if (!bpmnString || bpmnString.length === 0) {
      return;
    }

    const modeler = navigated
      ? new NavigatedViewer({
          container: containerRef.current as HTMLDivElement,
        })
      : new Viewer({
          container: containerRef.current as HTMLDivElement,
        });

    async function openDiagram(xmlString: string) {
      try {
        await modeler.importXML(xmlString);
      } catch (err) {
        console.error(err);
      }
    }
    setModeler(modeler);
    openDiagram(bpmnString);

    return () => {
      modeler.destroy();
    };
  }, [bpmnString, navigated]);

  useEffect(() => {
    if (modeler) {
      try {
        colorOptions?.forEach(({ elementId, intent }) => {
          setBpmnMarker(modeler, elementId, intent);
        });
      } catch (err) {
        console.error(err);
      }
    }
  }, [colorOptions, modeler]);

  const handleResetView = useCallback(() => {
    if (!modeler) {
      return;
    }
    setCurrentZoom(1);
    const canvas = modeler.get("canvas") as {
      viewbox(): { inner: ViewBoxInner };
      zoom: (mode: string, center: { x: number; y: number }) => void;
    };

    const { inner } = canvas.viewbox();

    const center = {
      x: inner.x + inner.width / 2,
      y: inner.y + inner.height / 2,
    };
    // zoom to fit full viewport
    canvas.zoom("fit-viewport", center);
  }, [modeler]);

  const handleZoom = (zoom: number) => () => {
    if (!modeler) {
      return;
    }
    setCurrentZoom(zoom);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (modeler.get("canvas") as any).zoom(zoom);
  };

  const setBpmnMarker = (
    modeler: Viewer,
    elementId: string,
    intent: string,
  ) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const canvas = modeler.get("canvas") as any;
    switch (intent) {
      case "ELEMENT_ACTIVATING":
      case "ELEMENT_ACTIVATED":
        canvas.addMarker(elementId, "activated");
        break;
      case "ELEMENT_COMPLETING":
      case "ELEMENT_COMPLETED":
      case "INCIDENT_RESOLVED":
        canvas.addMarker(elementId, "completed");
        break;
      case "ELEMENT_TERMINATING":
      case "ELEMENT_TERMINATED":
        canvas.addMarker(elementId, "terminated");
        break;
      case "INCIDENT_CREATED":
        canvas.addMarker(elementId, "incident");
        break;
      case "SEQUENCE_FLOW_TAKEN":
        canvas.addMarker(elementId, "flow_taken");
    }
  };

  useEffect(() => {
    if (width && width < 400) {
      return;
    }
    handleResetView();
  }, [width, handleResetView]);

  return (
    <div
      ref={containerRef}
      className="relative h-full w-full select-none"
      id="canvas"
      data-testid="canvas"
    >
      {control ? (
        <div className="absolute right-2 top-2 z-50 flex w-10 flex-col gap-2">
          <Button
            variant="secondary"
            helperText="Zoom In"
            onClick={handleZoom(currentZoom * (1 + zoomStrength))}
          >
            <ZoomInIcon sx={{ fontSize: "20px" }} />
          </Button>
          <Button
            variant="secondary"
            helperText="Zoom Out"
            onClick={handleZoom(currentZoom * (1 - zoomStrength))}
          >
            <ZoomOutIcon sx={{ fontSize: "20px" }} />
          </Button>
          <Button
            variant="secondary"
            helperText="Reset View"
            onClick={handleResetView}
          >
            <CenterFocusWeakIcon sx={{ fontSize: "20px" }} />
          </Button>
        </div>
      ) : null}
    </div>
  );
}

export function ResponsiveBpmnViewer({
  className,
  ...props
}: BpmnViewerProps & { className?: string }) {
  const [squareRef, { width, height }] = useElementSize();

  return (
    <div ref={squareRef} className={className}>
      <BpmnViewer width={width} height={height} {...props} />
    </div>
  );
}
