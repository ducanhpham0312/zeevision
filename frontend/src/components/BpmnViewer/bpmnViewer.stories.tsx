import type { Meta, StoryObj } from "@storybook/react";

import { BpmnViewer } from ".";
import { useEffect, useState } from "react";
import bpmnUrl from "../../bpmn/money-loan.bpmn";

const meta: Meta<typeof BpmnViewer> = {
  title: "Assets/BpmnViewer",
  component: BpmnViewer,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;

type Story = StoryObj<typeof meta>;

const LoadedBpmnViewer = () => {
  const [bpmnXml, setBpmnXml] = useState("");

  useEffect(() => {
    fetch(bpmnUrl)
      .then((res) => res.text())
      .then((bpmn) => setBpmnXml(bpmn));
  }, []);

  return (
    <div
      style={{
        border: "1px solid black",
        display: "flex",
        justifyContent: "center",
      }}
    >
      <BpmnViewer width={500} bpmnString={bpmnXml} />
    </div>
  );
};

export const Primary: Story = {
  render: () => <LoadedBpmnViewer />,
};
