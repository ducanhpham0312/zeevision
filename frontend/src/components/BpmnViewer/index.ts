export { BpmnViewer, ResponsiveBpmnViewer } from "./BpmnViewer";
