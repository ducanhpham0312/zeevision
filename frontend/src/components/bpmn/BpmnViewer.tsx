import Modeler from "bpmn-js/lib/Viewer";
import "./bpmn-js.css"
import { useEffect, useRef } from "react";

// create a modeler

function BpmnViewer({ bpmn, width }: { bpmn: string; width: number }) {
  const containerRef = useRef<HTMLDivElement>();

  useEffect(() => {
    const modeler = new Modeler({
      container: containerRef.current,
      width: width,
    });

    async function openDiagram(xml: any) {
      try {
        await modeler.importXML(xml);

        // access viewer components
        var canvas = modeler.get("canvas");

        // zoom to fit full viewport
        canvas.zoom("fit-viewport");

        containerRef.current?.classList.add("with-error");
        containerRef.current?.classList.remove("with-diagram");
      } catch (err) {
        containerRef.current?.classList.remove("with-error");
        containerRef.current?.classList.add("with-diagram");

        console.error(err);
      }
    }

    openDiagram(bpmn);

    return () => {
      modeler.destroy();
    };
  }, []);

  return <div ref={containerRef} style={{ userSelect: "none" }} id="canvas" />;
}

export default BpmnViewer;
