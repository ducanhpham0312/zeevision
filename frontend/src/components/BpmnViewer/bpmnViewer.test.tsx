import "@testing-library/jest-dom";
import { render, waitFor } from "@testing-library/react";
import NavigatedViewer from "bpmn-js/lib/NavigatedViewer";
import { BpmnViewer } from "./BpmnViewer";
import Viewer from "bpmn-js/lib/Viewer";

// Mock class constructor
jest.mock("bpmn-js/lib/NavigatedViewer", () => {
  return {
    __esModule: true, // ESModule compatibility
    default: jest.fn().mockImplementation(() => {
      return {
        // Mock methods and properties
        importXML: jest.fn(),
        destroy: jest.fn(),
        get: jest.fn().mockReturnValue({
          viewbox: () => ({ inner: {}, outer: {} }),
          zoom: jest.fn(),
        }),
      };
    }),
  };
});

jest.mock("bpmn-js/lib/Viewer", () => {
  return {
    __esModule: true, // ESModule compatibility
    default: jest.fn().mockImplementation(() => {
      return {
        // Mock methods and properties
        importXML: jest.fn(),
        destroy: jest.fn(),
        get: jest.fn().mockReturnValue({
          viewbox: () => ({ inner: {}, outer: {} }),
          zoom: jest.fn(),
        }),
      };
    }),
  };
});

describe("BpmnViewer Component", () => {
  it("correctly render the snapshot", () => {
    const bpmnString = "some-bpmn-xml-string";
    const { asFragment } = render(
      <BpmnViewer bpmnString={bpmnString} navigated={true} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders and initializes bpmn-js viewer", async () => {
    const bpmnString = "some-bpmn-xml-string";

    // Render the component
    const { getByTestId } = render(
      <BpmnViewer bpmnString={bpmnString} navigated={true} />,
    );

    await waitFor(() => {
      expect(getByTestId("canvas")).toBeInTheDocument();
    });
  });

  it("uses NavigatedViewer when navigated prop is true", async () => {
    const bpmnString = "some-bpmn-xml-string";
    render(<BpmnViewer bpmnString={bpmnString} navigated={true} />);

    await waitFor(() => {
      expect(NavigatedViewer).toHaveBeenCalled();
    });
  });

  it("uses Viewer when navigated prop is false", async () => {
    const bpmnString = "some-bpmn-xml-string";
    render(<BpmnViewer bpmnString={bpmnString} navigated={false} />);

    await waitFor(() => {
      expect(Viewer).toHaveBeenCalled();
    });
  });
});
