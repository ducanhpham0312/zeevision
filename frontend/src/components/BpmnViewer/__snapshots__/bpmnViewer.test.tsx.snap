// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpmnViewer Component correctly render the snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative h-full w-full select-none"
    data-testid="canvas"
    id="canvas"
  />
</DocumentFragment>
`;

exports[`ResponsiveBpmnViewer correctly render the ResponsiveBpmnViewer snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="relative h-full w-full select-none"
      data-testid="canvas"
      id="canvas"
    >
      <div
        class="absolute right-2 top-2 z-50 flex w-10 flex-col gap-2"
      >
        <button
          class="group hover:shadow-lg transition relative box-border text-accent hover:bg-hover active:bg-active rounded bg-second-accent px-3 py-2"
        >
          <div
            class="absolute -left-3 top-[50%] -translate-x-[100%] -translate-y-[50%] scale-50 rounded bg-[#202020] px-2 py-1 text-white opacity-0 transition group-hover:scale-100 group-hover:opacity-100"
          >
            <p
              class="whitespace-nowrap text-sm"
            >
              Zoom In
            </p>
          </div>
          <div
            class="transition-width absolute bottom-0 left-0 w-full transform rounded bg-accent duration-100"
            style="height: 0px;"
          />
          <span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1lrr7yx-MuiSvgIcon-root"
              data-testid="ZoomInIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
              <path
                d="M12 10h-2v2H9v-2H7V9h2V7h1v2h2v1z"
              />
            </svg>
          </span>
        </button>
        <button
          class="group hover:shadow-lg transition relative box-border text-accent hover:bg-hover active:bg-active rounded bg-second-accent px-3 py-2"
        >
          <div
            class="absolute -left-3 top-[50%] -translate-x-[100%] -translate-y-[50%] scale-50 rounded bg-[#202020] px-2 py-1 text-white opacity-0 transition group-hover:scale-100 group-hover:opacity-100"
          >
            <p
              class="whitespace-nowrap text-sm"
            >
              Zoom Out
            </p>
          </div>
          <div
            class="transition-width absolute bottom-0 left-0 w-full transform rounded bg-accent duration-100"
            style="height: 0px;"
          />
          <span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1lrr7yx-MuiSvgIcon-root"
              data-testid="ZoomOutIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14zM7 9h5v1H7z"
              />
            </svg>
          </span>
        </button>
        <button
          class="group hover:shadow-lg transition relative box-border text-accent hover:bg-hover active:bg-active rounded bg-second-accent px-3 py-2"
        >
          <div
            class="absolute -left-3 top-[50%] -translate-x-[100%] -translate-y-[50%] scale-50 rounded bg-[#202020] px-2 py-1 text-white opacity-0 transition group-hover:scale-100 group-hover:opacity-100"
          >
            <p
              class="whitespace-nowrap text-sm"
            >
              Reset View
            </p>
          </div>
          <div
            class="transition-width absolute bottom-0 left-0 w-full transform rounded bg-accent duration-100"
            style="height: 0px;"
          />
          <span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1lrr7yx-MuiSvgIcon-root"
              data-testid="CenterFocusWeakIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 15H3v4c0 1.1.9 2 2 2h4v-2H5v-4zM5 5h4V3H5c-1.1 0-2 .9-2 2v4h2V5zm14-2h-4v2h4v4h2V5c0-1.1-.9-2-2-2zm0 16h-4v2h4c1.1 0 2-.9 2-2v-4h-2v4zM12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 6c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
